design/batching.md

```diff
@@ -0,0 +1,944 @@
+                        Field { name: "time", data_type: Int64, is_nullable: false, metadata: {} },
+                    ]),
+                    is_nullable: true,
+                    metadata: {},
+                },
+            ),
+            is_nullable: false,
+            metadata: {},
+        },
+        Field {
+            name: "components",
+            data_type: Struct([
+                Field {
+                    name: "text_entry",
+                    data_type: List(
+                        Field {
+                            name: "item",
+                            data_type: Struct([
+                                Field { name: "body", data_type: Utf8, is_nullable: false, metadata: {} },
+                                Field { name: "level", data_type: Utf8, is_nullable: true, metadata: {} },
+                            ]),
+                            is_nullable: true,
+                            metadata: {
+                                "rerun.kind": "component".
+                                "rerun.component": "rerun.text_entry",
+                            },
+                        },
+                    ),
+                    is_nullable: false,
+                    metadata: {},
+                },
+            ]),
+            is_nullable: false,
+            metadata: {},
+        },
+    ],
+    metadata: {
+        "rerun.batch_id": "<BatchId>", 
+    },
+}
+```
+
+The one major difference here is that `event_id`, `entity_path` and `num_instances` join `timepoints` in having dedicated, top-level columns.
+This is important as, like timepoints, these will have to be handled separately (and deserialized into native types!) in order to drive some of the logic in the store.
+
+Lastly, we inject the `BatchId` as metadata.
+`BatchId` isn't used for any logic yet, but comes in very handy for debug purposes.
+
+At this point we might want to sort the batch by `(event_id, entity_path)`, which will greatly improve data locality once it sits in the store (see storage section below).
+
+That's also an opportunity to pre-compact the data: if two rows share the same timepoints with different components, we could potentially merge them together... that's a bit more controversial though as it means either dropping some `EventId`s, or supporting multiple `EventId`s for a single event.
+
+One last thing that needs to be taken care of before actually sending the data is compression / dictionary-encoding of some kind.
+Even some trivial gzipping should do wonders here.  
+TODO(cmc): do we already have some kind of compression in place? zstd should already be set up i figure?
+
+### Storage
+
+One of the major change storage-wise is the complete removal of component tables: index tables now reference the arrow data directly.  
+With the new design, the arrow buffers now store multiple rows of data. To reference a specific row, each index row must point to _a unit-length slice_ in a shared batch of arrow data.
```

Comment:
I don't understand the unit-length part here. That's only true if there is only one instance of the component?

If I log multiple point-clouds after each other into the same batch, I would get a single `point` component buffer, and each row would reference a slice of the buffer. It would only be unit-length if there was exactly one point in the point-cloud.

Reply:
It all depends at which layer you're slicing and how the data is actually laid out I guess, but the point I was trying to emphasize here is that we're slicing a _single row_ out of the many rows in the component column for the entire batch.

I'm honestly not sure how that's going to be laid out exactly once everything's said and done.